File: C++/40. Combination Sum II.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<int> temp;
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> vec;
        solve(0, candidates, temp, target, vec);
        vector<vector<int>> res;
        for(int i = 0; i<vec.size();i++){
            res.push_back(vec[i]);
        }
        return res;
    }
    void solve(int idx, vector<int>& candidates, vector<int>& temp, int target, vector<vector<int>>& vec){
        if(target == 0) {
            vec.push_back(temp);
            return;
        } 
        if(target<0){return;}
        for(int i=idx; i<candidates.size(); i++){
            if(i>idx && candidates[i] == candidates[i-1]){
                continue;
            }
            temp.push_back(candidates[i]);
            solve(i+1, candidates, temp, target-candidates[i], vec);
            temp.pop_back();
        }
    }
};

File: C#/40. Combination Sum II.cs
public class Solution {
    public IList<IList<int>> CombinationSum2(int[] candidates, int target) {
        List<int> temp = new();
        Array.Sort(candidates);
        List<IList<int>> list = new();
        solve(0, candidates, temp, target, list);
        List<IList<int>> res = new();
        for(int i = 0; i<list.Count;i++){
            res.Add(list[i]);
        }
        return res;
    }
    public void solve(int idx, int[] candidates, List<int> temp, int target, List<IList<int>> list){
        if(target == 0) {
            list.Add(new List<int>(temp));
            return;
        } 
        if(target<0){return;}
        for(int i=idx; i<candidates.Length; i++){
            if(i>idx && candidates[i] == candidates[i-1]){
                continue;
            }
            temp.Add(candidates[i]);
            solve(i+1, candidates, temp, target-candidates[i], list);
            temp.RemoveAt(temp.Count-1);
        }
    }
}
